### ISSUE TITLE
Enable Auto Backup Option Under Settings Menu (on by default)

### ISSUE THREAD

**Issue by ajbufort** [mit-cml/appinventor-sources#1922]
The current Settings menu in the top toolbar does not have an option to _automatically save (export for download) the current project (or all projects) and associated keystore at regular intervals_.  

While students (and **everyone**) should definitely get in the habit of manually backing things up regularly, they ought not be penalized (at least too heavily) for failing to do so.  I propose such a setting be added to App Inventor, and that it be on by default.  The interval could either be manually adjusted, or else set at a fixed interval.

I have, in fact, already implemented this in my local AI2 build as a fixed-interval feature.  Until discussed and approved here, however, I will not submit a pull request.

The files I had to touch were:

(under **appengine/src/com/google/appinventor/client...**)

```
Ode.java
OdeMessages.java
settings/user/GeneralSettings.java
TopToolbar.java
```

(under **appengine/src/com/google/appinventor/server/project/youngandroid**...)

```
YoungAndroidProjectService.java
```

(under **appengine/src/com/google/appinventor/shared/settings**...)

```
SettingsConstants.java
```

-Tony

![auto_save_option](https://user-images.githubusercontent.com/16006236/69673551-628ba080-104f-11ea-9cd0-1caa73246ebf.png)

**Comment by jisqyv** [comment]
@ajbufort What is the user experience when the interval is up? Does it just pop up a download box in the browser?

**Comment by ajbufort** [comment]
@jisqyv Yeah.  I know it is not as seamless as it could be, and I debated going the full nine on my first go and having a settings wizard in order to control interval and location of saves so one is not repeatedly prompted, but in the end I left it that way (with the pop-up) for now, and with the interval as a presumed constant.  If the user is away from the page and then returns after the timer task would have had occasion to execute multiple times, my limited experience using what I've set up has revealed that the task only executes once, waiting for the user to come back and address the dialog, and does not queue up multiple dialog boxes for each interval that has passed while the user was away.

**Comment by ewpatton** [comment]
> having a settings wizard in order to control [...] location of saves so one is not repeatedly prompted

I don't see how this is possible. Maybe in a local instance wrapped with Electron it could work, but browser settings and sandboxing would effectively make this impossible since it is the user settings that control whether the download dialog would appear and you don't usually have direct access to the user's file system. Do you have a proof of concept that works from within an App Engine instance?

> *save (export for download) the current project (or all projects) and associated keystore at regular intervals*

I am also concerned about the load that this would put on the App Inventor backend. In an offline instance it's a 1-1 correspondence between user and server, but during peak hours when we have thousands of people using App Inventor at a time (and typically in a 45 or 50 minute class window), either the backups wouldn't trigger at all (at 60 minutes), or we'd suddenly get a request for thousands of backups about the same time. Also, once people get beyond a certain number of projects the export all feature tends to fail due to the hard 1-minute timeout that Google enforces on all front-end requests. The export all feature would need to be rearchitected before this would be feasible in production (AFAIK this isn't enforced in the dev server, which is what most offline tools use).

Furthermore, I would be worried about the possible security weakness introduced on shared computers. For example, if this is on by default as suggested and I work on my App Inventor projects on a shared machine and backups include my keystore, then anyone else who uses that machine might be able to locate the backups and appropriate my keystore for their own use.

**Comment by ajbufort** [comment]
The thing that bugs me most regarding your concerns is your second one:

> either the backups wouldn't trigger at all (at 60 minutes), or we'd suddenly get a request for thousands of backups about the same time. Also, once people get beyond a certain number of projects the export all feature tends to fail due to the hard 1-minute timeout that Google enforces on all front-end requests.

That's the big dealbreaker to me.  

As far as #1 goes, I suppose it would be up to the user then as to how they want to configure auto-save behavior in terms of a dialog or no.  I don't see that as a big deal because, if it always has to be manual via a dialog every time anyway, so what, right?  It's only a once-an-hour or more nuisance.  At least, that's how I'd see the interval going.  But some could want that interval to be more frequent, in which case the smaller the interval, the more annoying the pop-up would be.

As for #3 and the issue of security risk, normally I'd say that if someone is foolish enough to save their stuff locally, as opposed to on a USB key or portable drive, then that's what they get.  But we are talking about kids here, so I cut them slack, so yeah - point taken.

In the end, I think the load and failure-possibility concerns make this DOA.  It sure is handy to me doing local development, though. :)

**Comment by ewpatton** [comment]
> if someone is foolish enough to save their stuff locally, as opposed to on a USB key or portable drive, then that's what they get. 

At least on Safari, I believe the default behavior is just for downloads to go into Downloads and you're not prompted where to save, so it might not necessarily be a case of the person accepting and putting it in the wrong place. Another possibility is that in the case where people are used to getting and accepting the prompt, they might just blindly click OK "because that's what you do" rather than thinking it through.

One possibility we did explore, but we're not sure whether to put it into production, is the ability to auto-backup to a Google Drive folder. This doesn't require the interrupt that comes from the Download dialog, so it addressed that issue, but obviously wouldn't work in the offline scenario so doesn't address your particular need.

**Comment by ajbufort** [comment]
Ok, well given what you said regarding auto-backup to a Google Drive folder, i would support that, so these kids have at least some automatic protection.
